# Рынок заведений общественного питания Москвы

[Описание проекта](#0)

[1 Импортируем библиотеки и файлы](#1) 
- [1.1 Изучение данных](#1.1)
- [1.2 Подготовка данных](#1.2)

[2 Анализ данных](#2)
- [2.1 Исследование соотношения видов объектов общественного питания по количеству](#2.1)
- [2.2 Исследуем соотношение сетевых и несетевых заведений по количеству](#2.2)
- [2.3 Определим для какого вида объекта общественного питания характерно сетевое распространение](#2.3)
- [2.4 Характеристики сетевых заведений](#2.4)
- [2.5 Для каждого вида объекта общественного питания опишем среднее количество посадочных мест](#2.5)
- [2.6 Выделим в отдельный столбец информацию об улице из столбца `address`](#2.6)
- [2.7 Определим топ-10 улиц по количеству объектов общественного питания](#2.7)
- [2.8 Определим число улиц с одним объектом общественного питания](#2.8)
- [2.9 Отобразим распределение количества посадочных мест для улиц с большим количеством объектов общественного питания](#2.9)

[3 Общий вывод](#3)
- [3.1 Рекомендации](#3.1)

<a name="0"></a>
**Описание проекта**
Вы решили открыть небольшое кафе в Москве. Оно оригинальное — гостей должны обслуживать роботы. Проект многообещающий, но дорогой. Вместе с партнёрами вы решились обратиться к инвесторам. Их интересует текущее положение дел на рынке — сможете ли вы снискать популярность на долгое время, когда все зеваки насмотрятся на роботов-официантов?

Вы — гуру аналитики, и партнёры просят вас подготовить исследование рынка. У вас есть открытые данные о заведениях общественного питания в Москве.

**Инструкция по выполнению проекта**

**Шаг №1. Загрузите данные и подготовьте их к анализу**
Загрузите данные о заведениях общественного питания Москвы. Убедитесь, что тип данных в каждой колонке — правильный, а также отсутствуют пропущенные значения и дубликаты. При необходимости обработайте их.
Путь к файлу: `/datasets/rest_data.csv`

**№2. Анализ данных**
* Исследуйте соотношение видов объектов общественного питания по количеству. Постройте график.
* Исследуйте соотношение сетевых и несетевых заведений по количеству. Постройте график.
* Для какого вида объекта общественного питания характерно сетевое распространение?
* Что характерно для сетевых заведений: много заведений с небольшим числом посадочных мест в каждом или мало заведений с большим количеством посадочных мест?
* Для каждого вида объекта общественного питания опишите среднее количество посадочных мест. Какой вид предоставляет в среднем самое большое количество посадочных мест? Постройте графики.
* Выделите в отдельный столбец информацию об улице из столбца `address`.
* Постройте график топ-10 улиц по количеству объектов общественного питания. Воспользуйтесь внешней информацией и ответьте на вопрос — в каких районах Москвы находятся эти улицы?
* Найдите число улиц с одним объектом общественного питания. Воспользуйтесь внешней информацией и ответьте на вопрос — в каких районах Москвы находятся эти улицы?
* Посмотрите на распределение количества посадочных мест для улиц с большим количеством объектов общественного питания. Какие закономерности можно выявить?

Сделайте общий вывод и дайте рекомендации о виде заведения, количестве посадочных мест, а также районе расположения. Прокомментируйте возможность развития сети.

**Шаг №3. Подготовка презентации**
Подготовьте презентацию исследования для инвесторов. Для создания презентации используйте любой удобный инструмент, но отправить презентацию нужно обязательно в формате pdf. Приложите ссылку на презентацию в markdown-ячейке в формате:

`Презентация: <ссылка на облачное хранилище с презентацией> `

Следуйте принципам оформления из темы «Подготовка презентации».

**Оформление:** Задание выполните в Jupyter Notebook. Программный код заполните в ячейках типа code, текстовые пояснения — в ячейках типа markdown. Примените форматирование и заголовки.

**Описание данных**

Таблица `rest_data`:
* `id` — идентификатор объекта;
* `object_name` — название объекта общественного питания;
* `chain` — сетевой ресторан;
* `object_type` — тип объекта общественного питания;
* `address` — адрес;
* `number` — количество посадочных мест.

## Импортируем библиотеки и файлы
<a name="1"></a>

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.options.display.max_colwidth = 150
from pandas.plotting import register_matplotlib_converters
register_matplotlib_converters()

import numpy as np
import random

import plotly.express as px
import plotly.graph_objs as go
from plotly.offline import iplot
import plotly.graph_objects as go


from plotly import __version__
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot


import seaborn as sns
import matplotlib.pyplot as plt

import requests
##import StringIO
import re
import numpy as np
from collections import defaultdict

##print __version__

init_notebook_mode(connected=True)

from io import BytesIO
import re
import io
import requests

In [2]:
#data = pd.read_csv('/datasets/rest_data.csv')
data = pd.read_csv(r'C:\Users\skd23\Dropbox\git\yandex_practikum_da\8_catering_market\datasets\rest_data.csv')

### Изучение данных
<a name="1.1"></a>

In [3]:
display('______________info_____________________')
display(data.info())
display('______________describe_________________')
display(data.describe())
display('______________head_____________________')
display(data.head(10))
display('______________tail_____________________')
display(data.tail(10))
display('_____________isna______________________')
display(data.isna().sum())
display('_____________duplicated________________')
display(data.duplicated().sum())
display('__________типы объектов общественного питания_________________')
display(data['object_type'].value_counts())
display('__________названия объектов общественного питания_________________')
display(data['object_name'].value_counts())




'______________info_____________________'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15366 entries, 0 to 15365
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   id           15366 non-null  int64 
 1   object_name  15366 non-null  object
 2   chain        15366 non-null  object
 3   object_type  15366 non-null  object
 4   address      15366 non-null  object
 5   number       15366 non-null  int64 
dtypes: int64(2), object(4)
memory usage: 720.4+ KB


None

'______________describe_________________'

id        number
count   15366.000000  15366.000000
mean   119720.066901     59.547182
std     73036.130732     74.736833
min       838.000000      0.000000
25%     28524.000000     12.000000
50%    144974.500000     40.000000
75%    184262.250000     80.000000
max    223439.000000   1700.000000

'______________head_____________________'

id               object_name chain object_type  \
0  151635                   СМЕТАНА   нет        кафе   
1   77874                    Родник   нет        кафе   
2   24309           Кафе «Академия»   нет        кафе   
3   21894                ПИЦЦЕТОРИЯ    да        кафе   
4  119365    Кафе «Вишневая метель»   нет        кафе   
5   27429  СТОЛ. ПРИ ГОУ СОШ № 1051   нет    столовая   
6  148815                  Брусника    да        кафе   
7   20957               Буфет МТУСИ   нет    столовая   
8   20958               КПФ СЕМЬЯ-1   нет    столовая   
9   28858            Столовая МТУСИ   нет    столовая   

                                               address  number  
0           город Москва, улица Егора Абакумова, дом 9      48  
1    город Москва, улица Талалихина, дом 2/1, корпус 1      35  
2            город Москва, Абельмановская улица, дом 6      95  
3              город Москва, Абрамцевская улица, дом 1      40  
4    город Москва, Абрамцевская улица, дом 9, корпус 1      50  
5   город Москва, Абрамцевская улица, дом 15, корпус 1     240  
6        город Москва, переулок Сивцев Вражек, дом 6/2      10  
7  город Москва, Авиамоторная улица, дом 8, строение 1      90  
8  город Москва, Авиамоторная улица, дом 8, строение 1     150  
9  город Москва, Авиамоторная улица, дом 8, строение 1     120

'______________tail_____________________'

id            object_name chain object_type  \
15356  220618            Шоколадница    да        кафе   
15357  218692            Шоколадница    да        кафе   
15358  213724            Шоколадница    да        кафе   
15359  222077  Кофейня «Шоколадница»    да        кафе   
15360  219759            Шоколадница    да        кафе   
15361  208537            Шоколадница    да        кафе   
15362  209264            Шоколадница    да        кафе   
15363  209186            Шоколадница    да        кафе   
15364  221900            Шоколадница    да        кафе   
15365  222535            Шоколадница    да        кафе   

                                                                                             address  \
15356                                                город Москва, Митинская улица, дом 36, корпус 1   
15357                                                 город Москва, площадь Джавахарлала Неру, дом 1   
15358                                                        город Москва, Варшавское шоссе, дом 87Б   
15359                                                     город Москва, Кантемировская улица, дом 47   
15360                                                            город Москва, улица Вавилова, дом 3   
15361                                                   город Москва, 3-й Крутицкий переулок, дом 18   
15362                                                       город Москва, улица Земляной Вал, дом 33   
15363                                                       город Москва, улица Земляной Вал, дом 33   
15364  город Москва, поселение Московский, Киевское шоссе, 22-й километр, домовладение 4, строение 1   
15365                                                         город Москва, Ходынский бульвар, дом 4   

       number  
15356     100  
15357      30  
15358      54  
15359      72  
15360      36  
15361      50  
15362      10  
15363      20  
15364      36  
15365      10

'_____________isna______________________'

id             0
object_name    0
chain          0
object_type    0
address        0
number         0
dtype: int64

'_____________duplicated________________'

0

'__________типы объектов общественного питания_________________'

кафе                                 6099
столовая                             2587
ресторан                             2285
предприятие быстрого обслуживания    1923
бар                                   856
буфет                                 585
кафетерий                             398
закусочная                            360
магазин (отдел кулинарии)             273
Name: object_type, dtype: int64

'__________названия объектов общественного питания_________________'

Столовая                   267
Кафе                       236
Шаурма                     234
KFC                        155
Шоколадница                142
                          ... 
Столовая ТЦ «Энтузиаст»      1
Охунова У.Д.                 1
СОК ООО АВТОКОМБИНАТ 3       1
АВТОКОМБИНАТ 3               1
Кальянная «Мята Lounge»      1
Name: object_name, Length: 10393, dtype: int64

Заведения с 0 количеством посадочных мест:

In [4]:
display(data[data['number'] == 0]['object_type'].unique())
print('Количество заведений с 0 количеством посадочных мест:', data[data['number'] == 0]['object_type'].count())

array(['закусочная', 'предприятие быстрого обслуживания', 'столовая',
       'кафетерий', 'кафе', 'магазин (отдел кулинарии)', 'буфет', 'бар',
       'ресторан'], dtype=object)

Количество заведений с 0 количеством посадочных мест: 1621


Определлим % пропущенных значений по каждому из столбцов:

In [5]:
percent_missing = data.isnull().sum() * 100 / len(data)
missing_value_data = pd.DataFrame({'column_name': data.columns,
                                 'percent_missing': percent_missing})
missing_value_data.columns = ['название столбца', '% пропущенных']
round(missing_value_data.reset_index(), 1)

index название столбца  % пропущенных
0           id               id            0.0
1  object_name      object_name            0.0
2        chain            chain            0.0
3  object_type      object_type            0.0
4      address          address            0.0
5       number           number            0.0

<a name="1.2"></a>
**Вывод:** 
* данные содержатся в таблице 15366 строк и 6 столбцов
* дублей нет, пропусков нет.
* отсутствие посадочных мест в объекте общественного питания сочтем за еще не открывшиеся заведения.

<a name="2"></a>
### Подготовка данных

Приведем адреса объекта общественного питания к нижнему регистру:

In [6]:
data['object_name'] = data['object_name'].str.lower()

Приведем адреса кафе к нижнему регистру:

In [7]:
data['object_name'] = data['object_name'].str.lower()

<a name="2"></a>
## Анализ данных
### Исследование соотношения видов объектов общественного питания по количеству
<a name="2.1"></a>

Сгруппирует типы объектов общественного питания по их количеству:

In [8]:
object_type_count = data.groupby('object_type').agg({'object_name':'count'})\
                                               .reset_index().sort_values(by='object_name', ascending=False)
object_type_count.columns = ['object_type', 'object_count']
object_type_count

object_type  object_count
3                               кафе          6099
8                           столовая          2587
7                           ресторан          2285
6  предприятие быстрого обслуживания          1923
0                                бар           856
1                              буфет           585
4                          кафетерий           398
2                         закусочная           360
5          магазин (отдел кулинарии)           273

In [9]:
trace = go.Bar(
    x = object_type_count.object_type, 
    y = object_type_count.object_count,
    marker_color='#22a6b3'
)
layout = go.Layout(
    title={
        'text': 'Соотношение видов объектов общественного питания по количеству',
        'y':0.85,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'},
)

fig = go.Figure(data = [trace], layout = layout)
iplot(fig)

По количеству в Москве преобладают кафе, более 6 тысяч.

### Исследуем соотношение сетевых и несетевых заведений по количеству
<a name="2.2"></a>

In [59]:
chain_rating = data.groupby('chain').agg({'object_name': 'count'})

chain_rating['object_type'] = (data.groupby("chain").agg({'object_type':"count"})
         .apply(lambda x: 100*x/x.sum()).sort_values('object_type', ascending=False)).round(2)

chain_rating.columns = ['количество', '% от общего кол-ва']
chain_rating.reset_index()

chain  количество  % от общего кол-ва
0    да        2968               19.32
1   нет       12398               80.68

In [69]:
fig = go.Figure(data=go.Pie(labels=chain_rating['количество'].value_counts()
                            .reset_index()['index'], values=data['chain'].value_counts()))
fig.update_layout(
    title={
        'text':'Соотношение сетевых и несетевых заведений по количеству',
        'y':0.87,
        'x':0.5})
for trace in fig.data:
    trace['labels']=['Несетевые', 'Сетевые']
fig.show()

По количеству в Москве преобладают несетевые 80.68% заведения общественного питания.

### Определим  для какого вида объекта общественного питания характерно сетевое распространение
<a name="2.3"></a>

In [12]:
object_type_chain = data.pivot_table(index='object_type', columns='chain', values='object_name', aggfunc='count')\
                                                            .reset_index().sort_values(by='да', ascending=False)
object_type_chain.columns = ['тип объекта общепита', 'сетевой', 'несетевой']
object_type_chain

тип объекта общепита  сетевой  несетевой
3                               кафе     1396       4703
6  предприятие быстрого обслуживания      791       1132
7                           ресторан      544       1741
5          магазин (отдел кулинарии)       78        195
2                         закусочная       56        304
4                          кафетерий       52        346
0                                бар       37        819
1                              буфет       11        574
8                           столовая        3       2584

In [71]:
fig = go.Figure()
fig.add_trace(go.Bar(
    y=object_type_chain['сетевой'],
    x=object_type_chain['тип объекта общепита'],
    name='сетевой',
    marker_color='#f0932b'
))
fig.add_trace(go.Bar(
    y=object_type_chain['несетевой'],
    x=object_type_chain['тип объекта общепита'],
    name='несетевой',
    marker_color='#686de0'
))

# Here we modify the tickangle of the xaxis, resulting in rotated labels.
fig.update_layout(barmode='group', xaxis_tickangle=-45,
        title={
        'text': 'Распределение объектов общественного питания по типу',
        'y':0.85,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
fig.show()

Сетевое распределение характерно для кафе, предприятий быстрого обслуживания и ресторанов.

### Характеристики сетевых заведений
<a name="2.4"></a>

In [14]:
data_chain_da = data[data['chain']=='да']
data_chain_da.describe()

id       number
count    2968.000000  2968.000000
mean   118193.376348    52.919811
std     67537.024490    57.140644
min       838.000000     0.000000
25%     59214.250000    15.000000
50%    129973.500000    40.000000
75%    176827.750000    72.000000
max    223318.000000   580.000000

Медианное значение посадочных мест 40

In [15]:
query_less40 = (data.query('chain == "да" and number < 40')['object_type']).count()
print('Количест заведений с менее 40 посадочными местами:', query_less40)

Количест заведений с менее 40 посадочными местами: 1445


In [16]:
query_more40 = (data.query('chain == "да" and number >= 40')['object_type']).count()
print('Количест заведений с более 40 посадочными местами:', query_more40)

Количест заведений с более 40 посадочными местами: 1523


Построим гистограмму количества посадочных мест:

In [17]:
data_chain_da_fig = px.histogram(data_chain_da, x="number")


data_chain_da_fig.update_layout(xaxis_title='Количество посадочных мест', yaxis_title='Количество заведений', 
        title={
        'text': 'Распределение количества посадочных мест',
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'}, showlegend=True)


data_chain_da_fig.update_traces(marker=dict(color='#40407a'))

data_chain_da_fig.show()

Отобразим в виде диаграмм размаха распределение сетевых заведений по количеству посадочных мест

In [18]:
fig_chain_set_object_type_box = px.box(data_chain_da, x="object_type", y="number")
fig_chain_set_object_type_box.update_layout(xaxis_title='', yaxis_title='Количество посадочных мест', 
        title={
        'text': 'Распределение сетевых заведений по количеству посадочных мест',
        'y':0.93,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'}, showlegend=True)

fig_chain_set_object_type_box.update_traces(marker=dict(color='#40407a'))


fig_chain_set_object_type_box.show()

Распределим количество посадочных мест по категориям взвисимости от количества посадочных мест:

In [19]:
def cnt_number (row):
    if row <= 15:
        return 'посадочных мест до 15'
    if row <= 40:
        return 'посадочных мест до 40'
    if row <= 72:
        return 'посадочных мест до 72'
    if row <= 100:
        return 'посадочных мест до 100'
    if row > 72:
        return 'посадочных мест более 100'
data['group_cnt_number'] = data['number'].apply(cnt_number)
#data.head()

In [20]:
per_group_cnt_number = (data.groupby("group_cnt_number").agg({"id":"count"}).apply(lambda x:100 * x/x.sum())).round(2).reset_index()
per_group_cnt_number.columns = ['Категория', '% по категории']
per_group_cnt_number.sort_values('% по категории', ascending=False)

Категория  % по категории
2      посадочных мест до 15           27.98
3      посадочных мест до 40           26.30
4      посадочных мест до 72           18.63
0  посадочных мест более 100           15.33
1     посадочных мест до 100           11.77

Отобразим графически распределение количества посадочных мест по категориям взвисимости от количества посадочных мест:

In [72]:
trace = go.Bar(
    x = per_group_cnt_number['Категория'], 
    y = per_group_cnt_number['% по категории'],
    marker_color='#575fcf'
)
layout = go.Layout(
    title={
        'text': 'Количество посадочных мест по категориям в %',
        'y':0.85,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'},
)

fig = go.Figure(data = [trace], layout = layout)
iplot(fig)

В Москве больше заведений с небольшим числом посадочных мест (до 15). 

### Для каждого вида объекта общественного питания опишем среднее количество посадочных мест
<a name="2.5"></a>

In [22]:
number_mean = data.groupby('object_type').agg({'number':'mean'}).round(0).sort_values(by='number', ascending=False).reset_index()
number_mean.columns = ['тип_объекта', 'посадочных_мест']
number_mean

тип_объекта  посадочных_мест
0                           столовая            130.0
1                           ресторан             97.0
2                              буфет             51.0
3                                бар             43.0
4                               кафе             40.0
5  предприятие быстрого обслуживания             21.0
6                          кафетерий              9.0
7                         закусочная              8.0
8          магазин (отдел кулинарии)              6.0

In [74]:
trace = go.Bar(
    x = number_mean['тип_объекта'], 
    y = number_mean['посадочных_мест'],
    marker_color='#D06224'
)
layout = go.Layout(
    title={
        'text': 'Распределение сетевых заведений по среднему количеству посадочных мест',
        'y':0.85,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'},
)

fig = go.Figure(data = [trace], layout = layout)
iplot(fig)

В Москве в среднем по количеству посадочных мест лидируют:
- столовая 130 мест
- ресторан 97 мест 
- буфет 51 место

### Выделим в отдельный столбец информацию об улице из столбца `address`
<a name="2.6"></a>

Проведем ряд преобразований со столбцом `address`:

In [24]:
data['address']=data['address'].str.lower()
#data['address']=data['address'].replace(['\«', '\»', '\'', 'город москва, ', 'город зеленоград, '],'',regex=True)
data['address']=data['address'].replace(['\«', '\»', '\'', 'город москва, '], '', regex=True)
data['address']=data['address'].replace(['-', ' '], ' ', regex=True)
data.head()

id             object_name chain object_type  \
0  151635                 сметана   нет        кафе   
1   77874                  родник   нет        кафе   
2   24309         кафе «академия»   нет        кафе   
3   21894              пиццетория    да        кафе   
4  119365  кафе «вишневая метель»   нет        кафе   

                               address  number        group_cnt_number  
0         улица егора абакумова, дом 9      48   посадочных мест до 72  
1  улица талалихина, дом 2/1, корпус 1      35   посадочных мест до 40  
2          абельмановская улица, дом 6      95  посадочных мест до 100  
3            абрамцевская улица, дом 1      40   посадочных мест до 40  
4  абрамцевская улица, дом 9, корпус 1      50   посадочных мест до 72

Создадим в `data` колонку `street` содержащую очищенный адрес:

In [25]:
data['street'] = data['address'].apply(lambda x: x.split(',')[0])

In [26]:
data.head()

id             object_name chain object_type  \
0  151635                 сметана   нет        кафе   
1   77874                  родник   нет        кафе   
2   24309         кафе «академия»   нет        кафе   
3   21894              пиццетория    да        кафе   
4  119365  кафе «вишневая метель»   нет        кафе   

                               address  number        group_cnt_number  \
0         улица егора абакумова, дом 9      48   посадочных мест до 72   
1  улица талалихина, дом 2/1, корпус 1      35   посадочных мест до 40   
2          абельмановская улица, дом 6      95  посадочных мест до 100   
3            абрамцевская улица, дом 1      40   посадочных мест до 40   
4  абрамцевская улица, дом 9, корпус 1      50   посадочных мест до 72   

                  street  
0  улица егора абакумова  
1       улица талалихина  
2   абельмановская улица  
3     абрамцевская улица  
4     абрамцевская улица

In [27]:
data['street'].value_counts().to_frame().head(11)

street
город зеленоград           232
проспект мира              204
профсоюзная улица          183
ленинградский проспект     173
пресненская набережная     167
варшавское шоссе           162
ленинский проспект         148
поселение сосенское        139
проспект вернадского       132
кутузовский проспект       114
каширское шоссе            112

Количество уникальных адресов:

In [28]:
len(data['street'].unique())

1853

### Определим топ-10 улиц по количеству объектов общественного питания
<a name="2.7"></a>

In [29]:
top10_streets = data.groupby(['street']).agg({'object_type':'count'}).reset_index()
top10_streets.columns = ['улица', 'количество_объектов_общепита']
top10_streets.sort_values(by='количество_объектов_общепита', ascending=False).head(11)

улица  количество_объектов_общепита
482         город зеленоград                           232
1076           проспект мира                           204
1080       профсоюзная улица                           183
699   ленинградский проспект                           173
1036  пресненская набережная                           167
416         варшавское шоссе                           162
702       ленинский проспект                           148
1025     поселение сосенское                           139
1073    проспект вернадского                           132
690     кутузовский проспект                           114
615          каширское шоссе                           112

В топ-10 улиц попал город Зеленоград (один из 12 административных центров Москвы) из-за особенностей зонального деления. Город поделен на районы и мкрорайоны с номерами без улиц и с улицами.

Прверим Зеленоград, так ли это:

In [30]:
data[data['address'].str.startswith('город зеленоград')].sample(10)

id  \
10709   29479   
2420    84456   
3445    19658   
7242    29408   
8832    83306   
9101   162722   
7243    70967   
10837   23848   
7236    24024   
13686  211330   

                                                                              object_name  \
10709                                                          гбоу г. москвы школа №1557   
2420                                                          столовая «гбоу школа №1692»   
3445                                                                             ели пели   
7242                                   столовая в школе 2045 корп.2011 (ранее школа 2038)   
8832                                                                  кафетерий клуб миэт   
9101                                                                            кафетерий   
7243   государственное бюджетное общеобразовательное учреждение города москвы школа №1528   
10837                                                                          школа 1710   
7236                                                                          кафе пронто   
13686                                                                 пекарня «дон батон»   

      chain object_type                                   address  number  \
10709   нет    столовая              город зеленоград, корпус 507     100   
2420    нет    столовая             город зеленоград, корпус 1016     160   
3445    нет        кафе  город зеленоград, корпус 124, строение 1      40   
7242    нет    столовая             город зеленоград, корпус 2011     180   
8832    нет   кафетерий   город зеленоград, площадь шокина, дом 1      16   
9101    нет   кафетерий             город зеленоград, корпус 435а       3   
7243    нет    столовая              город зеленоград, корпус 864     100   
10837   нет    столовая             город зеленоград, корпус 428а     100   
7236     да        кафе             город зеленоград, корпус 1549     220   
13686   нет        кафе             город зеленоград, корпус 1106       7   

                group_cnt_number            street  
10709     посадочных мест до 100  город зеленоград  
2420   посадочных мест более 100  город зеленоград  
3445       посадочных мест до 40  город зеленоград  
7242   посадочных мест более 100  город зеленоград  
8832       посадочных мест до 40  город зеленоград  
9101       посадочных мест до 15  город зеленоград  
7243      посадочных мест до 100  город зеленоград  
10837     посадочных мест до 100  город зеленоград  
7236   посадочных мест более 100  город зеленоград  
13686      посадочных мест до 15  город зеленоград

Отобразим графически топ-10 улиц по количеству объектов общественного питания:

In [31]:
top10_streets_bar = top10_streets.sort_values(by='количество_объектов_общепита', ascending=False)[:11]

In [75]:
trace = go.Bar(
    x = top10_streets_bar['улица'], 
    y = top10_streets_bar['количество_объектов_общепита'],
    marker_color='#1289A7'
)
layout = go.Layout(
    title={
        'text': 'Топ-10 адресов по количеству объектов общественного питания',
        'y':0.85,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'},
)

fig = go.Figure(data = [trace], layout = layout)
iplot(fig)

Определим в каких районах Москвы находятся эти улицы:

Используем открытые данные:
https://opengovdata.ru/en/dataset/

Импортируем данные:

In [33]:
url = 'https://raw.githubusercontent.com/skd2314/datasets/main/msk_data.xlsx'
data2 = pd.read_excel(url, index_col=0)
data2.head(10)

address  \
0         город москва, варшавское шоссе, дом 150, строение 2   
1                      город москва, косинская улица, дом 26а   
2                  город москва, гороховский переулок, дом 21   
3          город москва, тагильская улица, дом 4, строение 36   
4   город москва, ленинградский проспект, дом 73а, строение 3   
5                      город москва, улица вострухина, дом 5а   
6  город москва, улица космонавта волкова, дом 10, строение 1   
7          город москва, балтийская улица, дом 8, строение 15   
8             город москва, улица головачёва, дом 3, корпус 2   
9     город москва, климентовский переулок, дом 1, строение 3   

                               adm_area                             district  \
0          южный административный округ  муниципальный округ чертаново южное   
1      восточный административный округ          муниципальный округ вешняки   
2    центральный административный округ        муниципальный округ басманный   
3      восточный административный округ     муниципальный округ метрогородок   
4       северный административный округ            муниципальный округ сокол   
5  юго-восточный административный округ        муниципальный округ рязанский   
6       северный административный округ       муниципальный округ войковский   
7       северный административный округ         муниципальный округ аэропорт   
8  юго-восточный административный округ          муниципальный округ люблино   
9    центральный административный округ    муниципальный округ замоскворечье   

                street_name  
0          варшавское шоссе  
1           косинская улица  
2      гороховский переулок  
3          тагильская улица  
4    ленинградский проспект  
5          улица вострухина  
6  улица космонавта волкова  
7          балтийская улица  
8          улица головачёва  
9    климентовский переулок

In [34]:
data2['address'] = data2['address'].str.lower()

data2['address']=data2['address'].replace(['\«', '\»', '\'', 'город москва, '] ,'', regex=True)
data2['address']=data2['address'].replace(['-', '  '], ' ', regex=True)

data2['street'] = data2['address'].apply(lambda x: x.split(',')[0])
data2.head(10)

address  \
0         варшавское шоссе, дом 150, строение 2   
1                      косинская улица, дом 26а   
2                  гороховский переулок, дом 21   
3          тагильская улица, дом 4, строение 36   
4   ленинградский проспект, дом 73а, строение 3   
5                      улица вострухина, дом 5а   
6  улица космонавта волкова, дом 10, строение 1   
7          балтийская улица, дом 8, строение 15   
8             улица головачёва, дом 3, корпус 2   
9     климентовский переулок, дом 1, строение 3   

                               adm_area                             district  \
0          южный административный округ  муниципальный округ чертаново южное   
1      восточный административный округ          муниципальный округ вешняки   
2    центральный административный округ        муниципальный округ басманный   
3      восточный административный округ     муниципальный округ метрогородок   
4       северный административный округ            муниципальный округ сокол   
5  юго-восточный административный округ        муниципальный округ рязанский   
6       северный административный округ       муниципальный округ войковский   
7       северный административный округ         муниципальный округ аэропорт   
8  юго-восточный административный округ          муниципальный округ люблино   
9    центральный административный округ    муниципальный округ замоскворечье   

                street_name                    street  
0          варшавское шоссе          варшавское шоссе  
1           косинская улица           косинская улица  
2      гороховский переулок      гороховский переулок  
3          тагильская улица          тагильская улица  
4    ленинградский проспект    ленинградский проспект  
5          улица вострухина          улица вострухина  
6  улица космонавта волкова  улица космонавта волкова  
7          балтийская улица          балтийская улица  
8          улица головачёва          улица головачёва  
9    климентовский переулок    климентовский переулок

Объединим `data` и `data2` по адресу:

In [35]:
data3 = data.merge(data2, how='inner', on='address')
data3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15293 entries, 0 to 15292
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   id                15293 non-null  int64 
 1   object_name       15293 non-null  object
 2   chain             15293 non-null  object
 3   object_type       15293 non-null  object
 4   address           15293 non-null  object
 5   number            15293 non-null  int64 
 6   group_cnt_number  15293 non-null  object
 7   street_x          15293 non-null  object
 8   adm_area          15293 non-null  object
 9   district          15293 non-null  object
 10  street_name       15053 non-null  object
 11  street_y          15293 non-null  object
dtypes: int64(2), object(10)
memory usage: 1.5+ MB


In [36]:
data3.sample(10)

id              object_name chain object_type  \
497     69445                без забот   нет        кафе   
3814   153670       студент (закрытое)   нет        кафе   
8669   146957                 лига пап   нет        кафе   
8665    85501                мартиника   нет        кафе   
3598    25873                   бутчер   нет    ресторан   
13906  175698                 суши вок    да  закусочная   
12395  170763               legendario   нет         бар   
12660  188085  gournest гнездо гурмана   нет    ресторан   
10704  167377            ресторан paul   нет    ресторан   
6564    27975        столовая баловень   нет    столовая   

                                            address  number  \
497           мясницкая улица, дом 24/7, строение 1     202   
3814          ленинский проспект, дом 6, строение 1      50   
8669   краснопролетарская улица, дом 16, строение 2      98   
8665                   улица маршала рыбалко, дом 2      24   
3598                  улица малая дмитровка, дом 20      92   
13906                          улица грекова, дом 4      10   
12395                   улица дмитриевского, дом 10      35   
12660          бродников переулок, дом 10, корпус 2      40   
10704                пресненская набережная, дом 10      25   
6564      высоковольтный проезд, дом 1, строение 49      80   

                group_cnt_number                  street_x  \
497    посадочных мест более 100           мясницкая улица   
3814       посадочных мест до 72        ленинский проспект   
8669      посадочных мест до 100  краснопролетарская улица   
8665       посадочных мест до 40     улица маршала рыбалко   
3598      посадочных мест до 100     улица малая дмитровка   
13906      посадочных мест до 15             улица грекова   
12395      посадочных мест до 40       улица дмитриевского   
12660      посадочных мест до 40        бродников переулок   
10704      посадочных мест до 40    пресненская набережная   
6564      посадочных мест до 100     высоковольтный проезд   

                                      adm_area  \
497         центральный административный округ   
3814        центральный административный округ   
8669        центральный административный округ   
8665    северо-западный административный округ   
3598        центральный административный округ   
13906  северо-восточный административный округ   
12395         восточный административный округ   
12660       центральный административный округ   
10704       центральный административный округ   
6564   северо-восточный административный округ   

                                      district               street_name  \
497              муниципальный округ басманный           мясницкая улица   
3814              муниципальный округ якиманка        ленинский проспект   
8669              муниципальный округ тверской  краснопролетарская улица   
8665                муниципальный округ щукино     улица маршала рыбалко   
3598              муниципальный округ тверской     улица малая дмитровка   
13906  муниципальный округ северное медведково             улица грекова   
12395     муниципальный округ косино-ухтомский       улица дмитриевского   
12660             муниципальный округ якиманка        бродников переулок   
10704          муниципальный округ пресненский    пресненская набережная   
6564              муниципальный округ отрадное     высоковольтный проезд   

                       street_y  
497             мясницкая улица  
3814         ленинский проспект  
8669   краснопролетарская улица  
8665      улица маршала рыбалко  
3598      улица малая дмитровка  
13906             улица грекова  
12395       улица дмитриевского  
12660        бродников переулок  
10704    пресненская набережная  
6564      высоковольтный проезд

Оставим в `data3` только необходимые данные, удалив дубликаты:

In [37]:
data3 = data3[['id','object_name', 'chain', 'object_type', 'address', 'number', \
               'group_cnt_number', 'adm_area', 'district', 'street_name']].drop_duplicates()
data3.head()

id             object_name chain object_type  \
0  151635                 сметана   нет        кафе   
1   77874                  родник   нет        кафе   
2   24309         кафе «академия»   нет        кафе   
3   21894              пиццетория    да        кафе   
4  119365  кафе «вишневая метель»   нет        кафе   

                               address  number        group_cnt_number  \
0         улица егора абакумова, дом 9      48   посадочных мест до 72   
1  улица талалихина, дом 2/1, корпус 1      35   посадочных мест до 40   
2          абельмановская улица, дом 6      95  посадочных мест до 100   
3            абрамцевская улица, дом 1      40   посадочных мест до 40   
4  абрамцевская улица, дом 9, корпус 1      50   посадочных мест до 72   

                                  adm_area                         district  \
0  северо-восточный административный округ  муниципальный округ ярославский   
1       центральный административный округ    муниципальный округ таганский   
2       центральный административный округ    муниципальный округ таганский   
3  северо-восточный административный округ    муниципальный округ лианозово   
4  северо-восточный административный округ    муниципальный округ лианозово   

             street_name  
0  улица егора абакумова  
1       улица талалихина  
2   абельмановская улица  
3     абрамцевская улица  
4     абрамцевская улица

Отобразим в каких районах Москвы находятся эти топ-10 улиц:

In [38]:
top_district = data3.groupby(['district']).agg({'street_name':'count'}).reset_index()
top_district.columns = ['район', 'количество_улиц']
top_district.sort_values(by='количество_улиц', ascending=False).head(10)

район  количество_улиц
103       муниципальный округ тверской              774
83     муниципальный округ пресненский              715
9        муниципальный округ басманный              631
28     муниципальный округ даниловский              463
32   муниципальный округ замоскворечье              384
110      муниципальный округ хамовники              369
60       муниципальный округ мещанский              350
102      муниципальный округ таганский              340
6            муниципальный округ арбат              268
113    муниципальный округ хорошевский              257

In [39]:
top10_district = top_district.sort_values(by='количество_улиц', ascending=False)[:10]

trace = go.Bar(
    x = top10_district['район'], 
    y = top10_district['количество_улиц'],
    marker_color='#eccc68'
)

layout = go.Layout(
    title={
                'text': 'топ-10 районов по количеству улиц',
        'y':0.85,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'},
)

fig = go.Figure(data = [trace], layout = layout)
iplot(fig)

Большинство топ-10 районов находятся в центре Москвы

### Определим число улиц с одним объектом общественного питания
<a name="2.8"></a>

In [40]:
#top10_streets[top10_streets['количество_объектов_общепита']==1].head()
object_type_one = (data3.groupby('street_name')['id'].agg('count').sort_values(ascending=True).reset_index())
object_type_one.columns=['street_name','count_object_type']
object_type_one.head(10)

street_name  count_object_type
0                  яхромская улица                  1
1          староалексеевская улица                  1
2                стандартная улица                  1
3                волоцкой переулок                  1
4                 сретенский тупик                  1
5     средняя калитниковская улица                  1
6  средний овчинниковский переулок                  1
7  средний кондратьевский переулок                  1
8            старомарьинское шоссе                  1
9                спортивный проезд                  1

In [41]:
print('Число улиц с одним объектом общественного питания:', len(object_type_one.query('count_object_type == 1')))

Число улиц с одним объектом общественного питания: 576


Предварительно создадим `data4` в который добавим число улиц с одним объектом общественного питания:

In [42]:
data4 = data3.merge(object_type_one, how='inner', on='street_name')
data4.head(5)

id  object_name chain object_type                              address  \
0  151635      сметана   нет        кафе         улица егора абакумова, дом 9   
1  193477  grill-house   нет  закусочная       улица егора абакумова, дом 14а   
2  196756      пекарня   нет   кафетерий       улица егора абакумова, дом 14а   
3   77874       родник   нет        кафе  улица талалихина, дом 2/1, корпус 1   
4   83942   кафе даиси   нет        кафе          улица талалихина, дом 31а/2   

   number        group_cnt_number                                 adm_area  \
0      48   посадочных мест до 72  северо-восточный административный округ   
1       0   посадочных мест до 15  северо-восточный административный округ   
2      12   посадочных мест до 15  северо-восточный административный округ   
3      35   посадочных мест до 40       центральный административный округ   
4      78  посадочных мест до 100       центральный административный округ   

                          district            street_name  count_object_type  
0  муниципальный округ ярославский  улица егора абакумова                  3  
1  муниципальный округ ярославский  улица егора абакумова                  3  
2  муниципальный округ ярославский  улица егора абакумова                  3  
3    муниципальный округ таганский       улица талалихина                  9  
4    муниципальный округ таганский       улица талалихина                  9

In [43]:
district_count_object_type = data4.query('count_object_type == 1')[['district', 'street_name']]#.to_frame()
district_count_object_type.reset_index()

index                               district               street_name
0      109              муниципальный округ арбат     филипповский переулок
1      381         муниципальный округ сокольники          бабаевская улица
2      422           муниципальный округ аэропорт   1-й балтийский переулок
3      495  муниципальный округ хорошево-мневники     улица демьяна бедного
4      709          муниципальный округ ростокино           будайская улица
..     ...                                    ...                       ...
571  15017           муниципальный округ якиманка  2-й кадашёвский переулок
572  15018                 городской округ троицк          пионерская улица
573  15021                 поселение "мосрентген"       мкад, 44-й километр
574  15024                  поселение десеновское   улица дмитрия рябинкина
575  15025                    поселение щаповское   улица евгения родионова

[576 rows x 3 columns]

In [44]:
district_object_type_street = district_count_object_type.merge(object_type_one, how='inner', on='street_name')
district_object_type_street.columns = ['район', 'улица', 'количество_объектов_общепита']
district_object_type_street.head(5)

район                    улица  \
0              муниципальный округ арбат    филипповский переулок   
1         муниципальный округ сокольники         бабаевская улица   
2           муниципальный округ аэропорт  1-й балтийский переулок   
3  муниципальный округ хорошево-мневники    улица демьяна бедного   
4          муниципальный округ ростокино          будайская улица   

   количество_объектов_общепита  
0                             1  
1                             1  
2                             1  
3                             1  
4                             1

Отобразим в каких районах Москвы находятся эти улицы:

In [45]:
sort_district_object_type_street = district_object_type_street\
        .groupby('район')['количество_объектов_общепита']\
        .agg('count').sort_values(ascending=False).to_frame().reset_index()
sort_district_object_type_street.head(10)

район  количество_объектов_общепита
0       муниципальный округ таганский                            27
1       муниципальный округ хамовники                            26
2       муниципальный округ басманный                            22
3        муниципальный округ тверской                            20
4     муниципальный округ пресненский                            19
5    муниципальный округ марьина роща                            17
6      муниципальный округ сокольники                            15
7       муниципальный округ мещанский                            13
8   муниципальный округ замоскворечье                            12
9  муниципальный округ соколиная гора                            12

Отобразим графически топ-10 районов с улицами по 1 объекту общепита:

In [46]:
top10_sort_district_object_type_street = sort_district_object_type_street.sort_values(by='количество_объектов_общепита', ascending=False)[:10]

trace = go.Bar(
    x = top10_sort_district_object_type_street['район'], 
    y = top10_sort_district_object_type_street['количество_объектов_общепита'],
    marker_color='#95afc0'
)

layout = go.Layout(
    title={
                'text': 'Распределение количества улиц (с 1 объектом общепита) по районам',
        'y':0.85,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'},
)

fig = go.Figure(data = [trace], layout = layout)
iplot(fig)

### Отобразим распределение количества посадочных мест для улиц с большим количеством объектов общественного питания
<a name="2.9"></a>

In [47]:
number_street = data4.groupby(['street_name', 'count_object_type']).agg({'number':'mean'}).round(0)\
                .reset_index().sort_values(by='count_object_type', ascending=False)#[:15]
number_street.columns = ['улица', 'кол-во_объектов_общепита', 'среднее_кол-во_мест']
number_street.head()

улица  кол-во_объектов_общепита  среднее_кол-во_мест
1121           проспект мира                       204                 63.0
1125       профсоюзная улица                       181                 47.0
742   ленинградский проспект                       171                 53.0
1077  пресненская набережная                       167                 46.0
437         варшавское шоссе                       161                 53.0

In [48]:
fig = px.scatter(number_street, 
                x="среднее_кол-во_мест", 
                y="кол-во_объектов_общепита",
                marginal_x="box",
#               marginal_x="histogram",
                marginal_y="box", 
#                trendline="", 
#                template=""
                )

fig.update_layout(xaxis_title='Среднее количество посадочных мест', yaxis_title='Количество объектов общепита', title={
        'text': 'Среднее количество посадочных мест по улицам с большим количеством объектов',
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'}, showlegend=False)

fig.update_traces(marker=dict(color='#cd6133'))
#fig.write_image('test1.png')
fig.show()

In [49]:
#number_street_10 = number_street.sort_values(by='улица', ascending=False)[:10]
#number_street_10_list = number_street_10['улица'].to_list()
#number_street_10_list

Для улиц с большим количеством объектов общественного питания характерны заведения с средним количеством посадочных мест.

## Общий вывод
<a name="3"></a>

- Для Москвы в подавляющем большинстве характерны объекты общественного питания "Кафе"
- Сетевые заведения составляют всего 19.32%, несетевые 80.68%
- Сетевое распределения характерно для "кафе", "предприятие быстрого обслуживания" и рестораны
- Среднее количество посадочных мест по всем заведениям равно 53, медианное значение посадочных мест равно 40. Лидируют по среднему количеству посадочных мест "столовые" (130) и "ресторан" (97). Количеств заведений с менее 40 посадочными местами: 1445, количеств заведений с более 40 посадочными местами: 1523
- В топ-10 улиц по количеству объектов общественного питания попали проспекты, шоссе и набережные, что объясняется их протяженностью и большим потоком людей
- Топ-10 улиц располагаются в центральных районах Москвы
- Число улиц с одним объектом общественного питания: 576. Как правило это удаленные от центра улицы, переулки, тупики и проезды.

### Рекомендации
<a name="3.1"></a>

- Для развития сети, в районах не входящих в в топ-10, общественного питания можно рекомендовать открывать "предприятия быстрого обслуживания", "кафе" и "столовые", т.к они всегда пользуются спросом. 
- Количество посадочных мест будет обусловлено арендованым помещением, наличием конкурентов, но не более средней величины характерной для выбранного типа заведения.

- Для развития сети, в районах входящих в топ-10, общественного питания можно рекомендовать открывать "кафе", вследствие их большой популярности. Для повышения конкурентности "кафе", необходимо их "уникальность" (тематическое направление), с количеством посадочных мест меньше среднего.

- Исходя из банальной эрудиции, запрос на экзотические hitech-технологии выше в центрах городов, где выше платежеспособность населенья.
- Чем дальше от центра, тем проще запросы посетителей, следовательно роботизация необходимо рассматривать в разрезе автоматизации, для максимального снижения издержек.

Презентация: 